### DISCUSSION TITLE
Unrecoverable error: ioerror in setdistillerparams

### DISCUSSION THREAD

**Discussion by 1127** [ImageMagick/ImageMagick6#330] [General]
### ImageMagick version

6.9.10-97

### Operating system

Linux

### Operating system, version and so on

Amazon Linux 2

### Description

Did a yum update recently and immediately started getting errors - something about ImageMagick and Ghostscript. Ghostscript wasn't installed ... strange. We downloaded and compiled Ghostscript and everything worked. Then we realized that yum had a ghostscript install so we installed that. ImageMagick broke. We uninstalled everything and reinstalled everything - still isn't working... yes we tried to re-complile ghostscript again - nothing will fix this so far.

Background: We use ImageMagick to convert PDFs to JPGs (preview a PDF image on a website). Has worked for years and suddenly stopped. We upgraded and downgraded ghostscript and ImageMagick serveral times and can NOT fix this problem. 

We can reproduce the problem from the command line using a simple pdf - chicken.pdf... the command we are running is:
**`convert chicken.pdf output.jpg` or `convert chicken.pdf output.png`**

[chicken.pdf](https://github.com/user-attachments/files/16055725/chicken.pdf)

**Output is:**
convert: Postscript delegate failed `[ghostscript library 95.40] -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=pngalpha' -dTextAlphaBits=4 -dGraphicsAlphaBits=4 '-r72x72'  '-sOutputFile=/tmp/magick-5049IV31F-0PMGUg%d' '-f/tmp/magick-5049I17ImZwstTgF' '-f/tmp/magick-5049ep8Sux28NaD3'': Unrecoverable error: ioerror in setdistillerparams
Operand stack:
    --nostringval--
 @ error/ghostscript-private.h/InvokeGhostscriptDelegate/171.
convert: no images defined `output.jpg' @ error/convert.c/ConvertImageCommand/3229.

We are willing to pay an expert to help us with this problem!

### Steps to Reproduce

convert chicken.pdf output.jpg

**Comment by fmw42** [comment]
Your PDF and command works fine for me on IM  6.9.13-12 on Mac OSX with Ghostscript 10.03.0

What is your Ghostscript version?

**Comment by 1127** [comment]
Thanks fmw42... we confirmed it works on other platforms - even a fresh install of Amazon Linux 2.... we broke something on our install - and don't know how to fix.

**Comment by fmw42** [comment]
What is your Ghostscript version?  Are you sure you have only one version? Is Ghostscript in your $PATH when you install Imagemagick?  Can you convert your PDF with Ghostscript alone?

**Comment by 1127** [comment]
GPL Ghostscript 9.54.0 (2021-03-30)
Interesting - the command `gs -sDEVICE=png16m -sOutputFile=chicken2.jpg chicken.pdf` does convert it to a jpg... no errors works fine. Should I just be using Ghostscript? The actual code that runs is PHP... which outputs the error shown above - thus I figured if I can fix the command line, the PHP script will also work.

![chicken2](https://github.com/ImageMagick/ImageMagick/assets/7570821/b2cdc826-28a3-4502-aa1a-bf337e2155e5)

**Comment by fmw42** [comment]
You can just use Ghostscript.  But Imagick or PHP Imagemagick should work. I suspect you are having conflicting installs with Imagemagick Ghostscript and what you have in your native PHP install of Imagick/Imagemagick.

**Comment by 1127** [comment]
I should have asked - how would I know if I have two installs of ghostscript? There is only /bin/gs.... there is a "ghostscript" in there but it is a link to "gs". We did compile GS manually and install it... I *think* we deleted the binary "gs"... and then installed it with yum. But I don't know if that means we actually uninstalled it.

Maybe I need a windows-like uninstaller for gs? :-)

**Comment by fmw42** [comment]
type gs or where gs might help.  Or you could search different directories with find /directory -name gs or perhaps Ghostscript.  I suspect you have tried to install gs but it may already exist from your PHP install.  You may also have two Imagemagick installs if you installed one separate from PHP.  Sorry, I am not an expert on Linux.

**Comment by dlemstra** [comment]
ImageMagick is linked with Ghostscript and that is why you are seeing: `[ghostscript library 95.40]`. I would first recommend you to upgrade Ghostscript. The version you are using contains a remote code execution vulnerability. This might also fix that error.

If that doesn't resolve the issue you could consider recompiling ImageMagick with `--with-gslib` to make sure you are always using the command line version of Ghostscript instead of the library calls.

**Comment by dlemstra** [comment]
I changed this issue into a discussion because this doesn't seem to be a bug inside ImageMagick.

**Comment by 1127** [comment]
Thanks for your help - we still haven't been able to fix this problem. Upgrading to the latest ghostscript gets us to a point where we have all packages handled by AWS and yum package manager except for one thing... actually doing that got us into this problem (instead of yum installing ghostscript.... we manually installed).... any other ideas anyone?

**Reply by dlemstra** [reply]
I gave an option earlier?

**Reply by 1127** [reply]
Sorry, I should have thanked you dlemstra... your options were to upgrade Ghostscript (yum, in AWS at least, doesn't have anything newer) or recompile ImageMagick (again installed with yum and not compiled by me).

We are on Amazon Linux 2 which is near (at?) EOL. We'll be upgradeing to Amazon Linux 2023, that will fix the problem - and probably allow for newer versions of software.
